pkg/resource/deploy/step.go

```diff
@@ -951,25 +951,30 @@ func (s *ReadStep) Skip() {
 // resource by reading its current state from its provider plugin. These steps are not issued by the step generator;
 // instead, they are issued by the deployment executor as the optional first step in deployment execution.
 type RefreshStep struct {
-	deployment *Deployment       // the deployment that produced this refresh
-	old        *resource.State   // the old resource state, if one exists for this urn
-	new        *resource.State   // the new resource state, to be used to query the provider
-	provider   plugin.Provider   // the optional provider to use.
-	diff       plugin.DiffResult // the diff between the cloud provider and the state file
+	deployment *Deployment                                // the deployment that produced this refresh
+	old        *resource.State                            // the old resource state, if one exists for this urn
+	new        *resource.State                            // the new resource state, to be used to query the provider
+	provider   plugin.Provider                            // the optional provider to use.
+	diff       plugin.DiffResult                          // the diff between the cloud provider and the state file
+	cts        *promise.CompletionSource[*resource.State] // the completion source to signal when the refresh is complete
 }
 
 // NewRefreshStep creates a new Refresh step.
-func NewRefreshStep(deployment *Deployment, old *resource.State) Step {
+func NewRefreshStep(deployment *Deployment, cts *promise.CompletionSource[*resource.State], old *resource.State) Step {
 	contract.Requiref(old != nil, "old", "must not be nil")
 
 	// NOTE: we set the new state to the old state by default so that we don't interpret step failures as deletes.
 	return &RefreshStep{
 		deployment: deployment,
 		old:        old,
 		new:        old,
+		cts:        cts,
 	}
 }
 
+// True if this is a modern refresh step that should be respected by the snapshot system.
+func (s *RefreshStep) Modern() bool { return s.cts != nil }
```

Comment:
This is a terrible name, please suggest something better

Reply:
ack yes I think that's a good idea, Persisted is probably the right name for it.